### PR TITLE
Enforce dependent scope parameterized return producer

### LINKED ISSUE #669: Enforce @Dependent scope for producer methods with parameterized type variable return types
#### Description:
According to the CDI specification, a producer method whose return type is a parameterized type with a type variable (e.g., List<T>, Optional<T>) must be declared with @Dependent scope. If such a method is annotated with any other scope (e.g., @ApplicationScoped, @RequestScoped), the container must treat it as a definition error. 

#### Specification:
https://jakarta.ee/specifications/cdi/3.0/jakarta-cdi-spec-3.0#producer_method:~:text=If%20the%20producer%20method%20return%20type%20is%20a%20parameterized%20type%20with,automatically%20detects%20the%20problem%20and%20treats%20it%20as%20a%20definition%20error.

https://jakarta.ee/specifications/cdi/3.0/jakarta-cdi-spec-3.0#producer_method:~:text=If%20a%20producer%20method%20return%20type%20is%20a%20type%20variable%20or%20an%20array%20type%20whose%20component%20type%20is%20a%20type%20variable%20the%20container%20automatically%20detects%20the%20problem%20and%20treats%20it%20as%20a%20definition%20error.

#### Type of language feature proposed:
_Select all that apply_
- [x] diagnostic
- [ ] quick-fix
- [ ] snippet
- [ ] other, please specify: 

#### Invalid Example: 
```
@ApplicationScoped 
public class MyProducer { 
    @Produces 
    public <T> List<T> produceList() { 
        return new ArrayList<>(); 
    } 
} 
//  Invalid: parameterized type with type variable must be scoped @Dependent 
```
#### Valid Example: 
```
@Dependent 
public class MyProducer { 
    @Produces 
    public <T> List<T> produceList() { 
        return new ArrayList<>(); 
    } 
} 
```
A producer method return type must not be a type variable or an array whose component type is a type variable.

invalid
```
@Dependent
public class InvalidProducerMethod {

    @Produces
    public <T> T produceBareType() {   // return type is just a type variable
        return null;
    }

    @Produces
    public <T> T[] produceBareArray() { //  array of type variable
        return null;
    }
}

```

Valid

```
@Dependent
public class ValidProducerMethod {

    @Produces
    public String produceString() {   //  concrete type
        return "Hello";
    }

    @Produces
    public List<String> produceList() { // parameterized with actual type
        return new ArrayList<>();
    }
}
```

### PR DESCRIPTION
**Pull request by Rejoice472**
fixes https://github.com/eclipse-lsp4jakarta/lsp4jakarta/issues/669

https://github.com/user-attachments/assets/7f4d6f4d-f0a0-4356-9242-1db228af7ba4

### CODE REVIEW

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 1]
```diff
@@ -62,73 +67,67 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
-                                                                 Messages.getMessage("InvalidWildcardTypeInProducerField"),
-                                                                 range,
-                                                                 Constants.DIAGNOSTIC_SOURCE,
-                                                                 null,
-                                                                 ErrorCode.InvalidWildcardTypeInProducerField,
-                                                                 DiagnosticSeverity.Error));
-                    }
+                } else if (DiagnosticUtils.getMatchedJavaElementNames(type, annotationNames,
+                                                                      new String[] { Constants.PRODUCES_FQ_NAME }).size() > 0) {
+                    checkProducerMember(context, uri, diagnostics, type, annotationNames,
```
Comment (archana-1924) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java:89]: If the message key and error code has same value, then we can pass just ErrorCodes and do ErrorCode.name() to get the string value for the purpose of message key

**Review by archana-1924 — CHANGES_REQUESTED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 2]
```diff
@@ -62,73 +67,67 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
+
             for (IField field : type.getFields()) {
                 String[] annotationNames = DiagnosticUtils.getAnnotationNames(field);
 
-                // Use if-else since @Inject and @Produces don't appear on the same field
-                if (hasAnnotation(type, annotationNames, Constants.INJECT_FQ_NAME)) {
-                    String typeSignature = field.getTypeSignature();
-                    if (containsWildcard(typeSignature)) {
-                        Range range = PositionUtils.toNameRange(field, context.getUtils());
+                if (DiagnosticUtils.getMatchedJavaElementNames(type, annotationNames,
```
Comment (archana-1924) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java]: If we just need boolean, can't we use isMatchedAnnotation(ICompilationUnit unit, IAnnotation[] annotations, String annotationFQName)?

We can pass field.getAnnotations()

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 3]
```diff
@@ -62,73 +67,67 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
-                        Range range = PositionUtils.toNameRange(field, context.getUtils());
-                        diagnostics.add(context.createDiagnostic(uri,
-                                                                 Messages.getMessage("InvalidWildcardTypeInProducerField"),
-                                                                 range,
-                                                                 Constants.DIAGNOSTIC_SOURCE,
-                                                                 null,
-                                                                 ErrorCode.InvalidWildcardTypeInProducerField,
-                                                                 DiagnosticSeverity.Error));
-                    }
+                } else if (DiagnosticUtils.getMatchedJavaElementNames(type, annotationNames,
```
Comment (archana-1924) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java]: Same here

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 4]
```diff
@@ -62,73 +67,67 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
+                                                                      new String[] { Constants.PRODUCES_FQ_NAME }).size() > 0) {
+                    checkProducerMember(context, uri, diagnostics, type, annotationNames,
+                                        method.getReturnType(), typeParamNames, scopeFQNames,
+                                        PositionUtils.toNameRange(method, context.getUtils()),
+                                        new ErrorCode[] { ErrorCode.InvalidWildcardTypeInProducerMethod,
+                                                          ErrorCode.InvalidProducerMethodWithBareTypeVariableReturnType,
+                                                          ErrorCode.InvalidProducerMethodWithTypeVariableAndNonDependentScope },
```
Comment (archana-1924) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java]: Pass Just ErrorCodes

**Review by archana-1924 — CHANGES_REQUESTED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 5]
```diff
@@ -137,79 +126,152 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
-        // Check for parameterized types
+    /**
+     * Returns {@code true} if {@code typeSignature} is a parameterized type that contains
+     * at least one type variable in its type arguments (recursively).
+     *
+     * @param typeSignature the JDT type signature to check
+     * @param typeParamNames the class-level declared type parameter names
+     * @return {@code true} if the signature is a parameterized type containing a type variable
+     */
+    private boolean containsTypeVariable(String typeSignature, Set<String> typeParamNames) {
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java:253]: Can you add examples to documentation?

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java [hunk 6]
```diff
@@ -137,79 +126,152 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
+     * Returns {@code true} if {@code typeSignature} contains a wildcard type parameter
+     * ({@code ?}, {@code ? extends}, or {@code ? super}) anywhere in the type tree.
      *
-     * @param typeSignature the type signature to check
-     * @return true if the signature is a wildcard, false otherwise
+     * @param typeSignature the JDT type signature to check
+     * @return {@code true} if the signature contains a wildcard
      */
-    private boolean isWildcardSignature(String typeSignature) {
+    private boolean containsWildcard(String typeSignature) {
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/cdi/CdiWildcardDiagnosticsParticipant.java:284]: Can you add examples to documentation?

**Review by Joseph-Bineesh — CHANGES_REQUESTED** [review]
Update documentation. Rest looks good.

**Review by Joseph-Bineesh — APPROVED** [review]
LGTM

**Comment by Joseph-Bineesh** [comment]
Please address the pipeline failure

**Review by archana-1924 — APPROVED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.test/src/main/java/org/eclipse/lsp4jakarta/jdt/test/cdi/CdiWildcardBeanTypesTest.java [hunk 7]
```diff
@@ -117,36 +117,36 @@ public void wildcardBeanTypes() throws Exception {
...
 
         Diagnostic injectMethodNestedWildcard = d(136, 55, 58,
                                                   "Wildcard types are not legal bean types. Injection method parameters must use concrete parameterized types without wildcards (?, ? extends, ? super).",
-                                                  DiagnosticSeverity.Error, "jakarta-cdi", "InvalidWildcardTypeInInjectField");
+                                                  DiagnosticSeverity.Error, "jakarta-cdi", "InvalidWildcardTypeInInjectMethod");
 
         Diagnostic injectMethodArrayWildcard = d(140, 43, 48,
                                                  "Wildcard types are not legal bean types. Injection method parameters must use concrete parameterized types without wildcards (?, ? extends, ? super).",
-                                                 DiagnosticSeverity.Error, "jakarta-cdi", "InvalidWildcardTypeInInjectField");
+                                                 DiagnosticSeverity.Error, "jakarta-cdi", "InvalidWildcardTypeInInjectMethod");
```
Comment (venmanyarun) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.test/src/main/java/org/eclipse/lsp4jakarta/jdt/test/cdi/CdiWildcardBeanTypesTest.java:140]: Add a test in both lsp4jakarta and LTI for negative Tests with zero diagnostcis

**Review by venmanyarun — CHANGES_REQUESTED** [review]

**Review by venmanyarun — APPROVED** [review]
